Allow custom namespace creation in e2e framework

mfojtik opened the pull request.

alex-mohr commented:
@k8s-bot ok to test

ixdy commented:
what's the motivation for adding this functionality?

mfojtik commented:
@ixdy allow to use OpenShift 'Project' resource instead of the 'namespace'.

ixdy commented on test/e2e/framework.go:
```diff
@@ -38,13 +38,31 @@ type Framework struct {
+// NewCustomFramework makes a new framework and sets up a BeforeEach/AfterEach
+// for you. It also allows you to customize namespace creation by supplying a
+// custom function.
+func NewCustomFramework(baseName string, nsCreateFunc func(string, *client.Client) (*api.Namespace, error)) *Framework {
```
somewhat minor nit, but the name "NewCustomFramework" doesn't really describe well how it's different from NewFramework. In particular, what's "Custom"? The name implies nothing about the namespace creation function.

of course, I'm also having trouble trying to think of a much better name to recommend that isn't unnecessarily verbose.

ixdy commented on test/e2e/framework.go:
```diff
@@ -38,13 +38,31 @@ type Framework struct {
 
 // NewFramework makes a new framework and sets up a BeforeEach/AfterEach for
 // you (you can write additional before/after each functions).
 func NewFramework(baseName string) *Framework {
```
probably better to have this call the new function, passing `createTestingNS`, rather than duplicating the code.

mfojtik replied:
good catch, fixed.

mfojtik commented:
@ixdy sorry for the delay, I was PTO.

I fixed all your suggestions and I renamed the function to `InitializeFramework` (not a great name but I think better than `NewCustomFramework` ;-)

mfojtik commented:
@deads2k FYI

ixdy commented:
This PR seems to have pulled in other changes? Bad rebase?

ixdy commented:
ping @mfojtik - are you still interested in getting this PR merged?

mfojtik commented:
Yes, will rebase tomorrow. Thanks for reminding me :)
On Aug 31, 2015 8:36 PM, "Jeff Grafton" <notifications@github.com> wrote:

> ping @mfojtik <https://github.com/mfojtik> - are you still interested in
> getting this PR merged?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/12221#issuecomment-136457438>
> .
>

ixdy commented:
ping @mfojtik

ixdy commented:
Assigning back to you until this is rebased. Please reasssign when complete.

ixdy commented:
er, I guess I can't do that.

soltysh commented:
@mfojtik ping we carrying patch for this, could you please handle this one to simplify our rebase process, pretty, pretty please :heart_eyes:

mfojtik commented:
@soltysh @ixdy sorry for this to take that long... I think we only need a special function to create the namespace with, no need to move helpers...

ixdy commented:
lgtm

wojtek-t commented:
@k8s-bot test this please

wojtek-t commented:
@k8s-bot test this please

mfojtik commented:
@wojtek-t i don't think the e2e failures are related to this, but I can be wrong :-)

wojtek-t commented:
@mfojtik - I agree - these seems like flakes

ixdy commented:
@k8s-bot test this please

ixdy commented:
Aha, this is due to your change:

```console
$ grep Framework{ test/e2e/*
cluster_upgrade.go:             f := Framework{BaseName: "cluster-upgrade"}
container_probe.go:     framework := Framework{BaseName: "container-probe"}
daemon_restart.go:      framework := Framework{BaseName: "daemonrestart"}
daemon_set.go:  f := &Framework{BaseName: "daemonsets"}
density.go:     framework := Framework{BaseName: "density", NamespaceDeletionTimeout: time.Hour}
etcd_failure.go:        framework := Framework{BaseName: "etcd-failure"}
framework.go:   f := &Framework{
latency.go:     framework := Framework{BaseName: "latency", NamespaceDeletionTimeout: time.Hour}
load.go:        framework := Framework{BaseName: "load", NamespaceDeletionTimeout: time.Hour}
resize_nodes.go:        framework := Framework{BaseName: "resize-nodes"}
scheduler_predicates.go:        framework := Framework{BaseName: "sched-pred"}
serviceloadbalancers.go:        framework := Framework{BaseName: "servicelb"}
```

All of these tests that are creating the `Framework` struct directly instead of calling `NewFramework` are incompatible with your change.

mfojtik commented:
@ixdy this seems to be fixed (all the test cases)

soltysh commented:
LGTM

soltysh commented:
@ixdy can we get a review here, please.

ixdy commented on test/e2e/framework.go:
```diff
@@ -44,13 +44,23 @@ type Framework struct {
+
+// InitializeFramework initialize the framework by allowing to pass a custom
+// namespace creation function.
+func InitializeFramework(baseName string, nsCreateFunc func(string, *client.Client) (*api.Namespace, error)) *Framework {
```
sorry to rework this again, but given the number of additional things added to the `Framework` struct since this PR was first written (NamespaceDeletionTimeout, GatherKubeSystemResourceUsageData, etc) I'm wondering if maybe you should just override `nsCreateFunc` when necessary, rather than having a separate initialization method.

mfojtik replied:
@ixdy sorry for this to take so long.... yeah you're right, a standalone func that allows to override the createFunc makes sense

mfojtik committed:
Allow to override the e2e namespace creation function

mfojtik commented:
@ixdy @smarterclayton I reworked this a little bit so:

We can now write our own initializer for framework and override the namespace initialization like:

```go
f := &Framework{...}
f.SetNsCreateFunc(myCustomFunc).InstallGinkgoHooks()
```

That means we can change the `createTestingNS` func to create "project" instead of "namespace" in OpenShift and Kubernetes E2E will stay untouched.

soltysh commented:
I like the new approach even better. I'm sustaining my lgtm.

ixdy commented on test/e2e/framework.go:
```diff
@@ -57,12 +59,23 @@ type Framework struct {
 
+// SetNsCreateFunc allows to override the default namespace creation function
+// This function must be called before InstallGinkgoHooks() to take effect.
+func (f *Framework) SetNsCreateFunc(fn func(string, *client.Client) (*api.Namespace, error)) *Framework {
```
a setter seems a little heavyweight for what this is doing - why not just set NsCreateFunc directly?

ixdy commented on test/e2e/framework.go:
```diff
@@ -57,12 +59,23 @@ type Framework struct {
+}
 
+// SetNsCreateFunc allows to override the default namespace creation function
+// This function must be called before InstallGinkgoHooks() to take effect.
```
if you use NewFramework (as we intend everyone to do), there's no way to set this before InstallGinkgoHooks is called.

ixdy replied:
actually, I don't think it's necessary that you call this before `InstallGinkgoHooks` is called. `f.beforeEach` and `f.afterEach` aren't called until Ginkgo starts running test cases, which will occur after all the initial framework initialization is done. (Calling Ginkgo's `BeforeEach` and `AfterEach` just registers those handlers.)

so doing
```go
f := NewFramework("foo")
f.NsCreateFunc = myfunc
```

should work as intended.

wojtek-t commented:
@mfojtik - what's the status of this PR? Can we close it?

mfojtik commented:
@wojtek-t I'm testing @ixdy approach right now, will come back to this PR shortly

liggitt commented:
@mfojtik bump

deads2k commented:
@mfojtik rebase on master for rebase.  Dropping for now.

soltysh commented:
I'll be picking this up, will create a new PR and post link here.

liggitt commented:
can you work on the hook to do things as part of namespace creation as well so we can inject policy updates in our e2e?

soltysh commented:
That's the plan.

ixdy commented:
Should we close this PR then?
